server/src/main/java/org/elasticsearch/common/util/BytesRefHash.java

```diff
@@ -147,6 +151,15 @@ public long find(BytesRef key) {
         return find(key, key.hashCode());
     }
 
+    /**
+     * Allows finding a key in the hash in a thread safe manner, by providing an intermediate
+     * BytesRef reference to storing intermediate results. As long as each thread provides
+     * its own intermediate instance, this method is thread safe.
+     */
+    public long threadSafeFind(BytesRef key, BytesRef intermediate) {
+        return find(key, key.hashCode(), intermediate);
+    }
```

Comment:
I was thinking something like this:
```
    private long threadSafeFind(BytesRef key, BytesRef intermediate) {
        return find(key, key.hashCode(), intermediate);
    }
    
    public Finder newFinder() {
        return new Finder();
    }

    public class Finder {
        private final BytesRef intermediate = new BytesRef();

        public long find(BytesRef key) {
            return threadSafeFind(key, intermediate);
        }
    }
```

Reply:
I don't think that is so bad. I think we should still inherit from `AbstractIpScriptFieldQuery`

Then:
```
 @Override
    protected final boolean matches(BytesRef[] values, int count) {
        throw new UnsupportedOperationException("should not be called, use createTwoPhaseIterator instead");
    }
```

``` @Override
    protected final TwoPhaseIterator createTwoPhaseIterator(IpFieldScript scriptContext, DocIdSetIterator approximation) {
```

For the twophase iterator

Reply:
I see - I'm trying that. I'll add some tests for the new Find class to ensure it is thread safe.